akka-remote/src/main/scala/akka/remote/artery/Codecs.scala

```diff
@@ -156,35 +137,32 @@ private[remote] class Decoder(
         val envelope = grab(in)
         envelope.parseHeader(headerBuilder)
 
-        // FIXME: Instead of using Strings, the headerBuilder should automatically return cached ActorRef instances
-        // in case of compression is enabled
-        // FIXME: Is localAddress really needed?
-
-        val sender =
-          if (headerBuilder.isNoSender)
-            OptionVal.None
-          else {
-            senderCache.get(headerBuilder.senderActorRef) match {
-              case null ⇒
-                val ref = resolveActorRefWithLocalAddress(headerBuilder.senderActorRef)
-                // FIXME this cache will be replaced by compression table
-                if (senderCache.size() >= 1000)
-                  senderCache.clear()
-                senderCache.put(headerBuilder.senderActorRef, ref)
-                OptionVal(ref)
-              case ref ⇒ OptionVal(ref)
-            }
-          }
-
-        val recipient =
-          if (headerBuilder.isNoRecipient)
-            OptionVal.None
-          else
-            resolveRecipient(headerBuilder.recipientActorRef)
-
         val originUid = headerBuilder.uid
         val association = inboundContext.association(originUid)
 
+        val recipient: OptionVal[InternalActorRef] = headerBuilder.recipientActorRef match {
+          case OptionVal.Some(ref) ⇒ OptionVal(ref.asInstanceOf[InternalActorRef])
+          case _                   ⇒ resolveRecipient(headerBuilder.recipientActorRefPath)
+        }
+
+        val sender: InternalActorRef = headerBuilder.senderActorRef match {
+          case OptionVal.Some(ref) ⇒ ref.asInstanceOf[InternalActorRef]
+          case _                   ⇒ resolveActorRefWithLocalAddress(headerBuilder.senderActorRefPath)
```

Comment:
ok, I see now.
`OptionVal.None` might be easier to read.
We might need some caching of the resolved actor refs also, in addition to what the compression table holds. resolve is super slow and we want to return the same `RemoteActorRef` instances as much as possible. Not sure what is best. Similar for recipient.

This is ok for this PR.